### ISSUE TITLE
Duplex problem with external USB interface in Linux

### ISSUE THREAD

**Issue by MarcBoule** [thestk/rtaudio#118]
Hi,

I'm using an application (VCVRack) that uses RtAudio as its sound engine, and would like to report a possible issue. When using the app in Ubuntu with a Steinberg UR22mkii usb audio interface, I get choppy sound. The app opens a stream with 2 input channels and 2 output channels, which corresponds to what the UR22mkii offers. The computer is a Lenovo T430 laptop. When I use the onboard audio, everything works well. Also, when I connect the UR22mkii to my macbook air and run the mac verison of the app in MacOS, all is perfect.

So I know that the hardware works, and the software also, but the UR22mkii/Ubuntu combination in particular is problematic. When I run the test programs that come with RtAudio, the _playraw_, _playsaw_, _record_, _audioprobe_ programs all work perfectly, but when I run _duplex_, I get choppy sound. If I modify _playsaw_ (which works) to also make it open 2 input channels (three extra lines of code), the output becomes choppy. I can collaborate to do more testing if needed, I realise the author of RtAudio can't own every USB audio interface out there :-)

Best regards,

Marc Boulé

EDIT: Test was using ALSA on Linux. Possible link with issue #89, although no seg faults on my side.

**Comment by garyscavone** [comment]
Did you try changing the bufferFrames size, perhaps doubling it? That is what controls the input / output latency.

**Comment by MarcBoule** [comment]
I tried many bufferFrames sizes ranging from 128 to 16384, and although the choppiness changes in nature, problems remain throughout, unfortunately.

**Comment by MarcBoule** [comment]
Also, I used 44100 sampling frequency, and repeated the tests with 1 channel instead of 2, and the problem remained, so it likely isn't related to interleaving.

**Comment by garyscavone** [comment]
Potentially sounds like a driver issue.

**Comment by MarcBoule** [comment]
What's interesting though, is that with a 16384 length buffer, I can hear a few clicks per second (likely 44100/16384 clicks per second), but the remaining audio seems ok. When I start a playsaw with my small code mod to open input channels to recreate the problem, there is an observable lag before the audio content starts, but the clicks start right away. Not sure what can be determined from this, but thought I'd mention it in case. If ever you have any ideas for further tests I could try, in order to zero-in on the problem, please let me know.

**Comment by spktkpkt** [comment]
@garyscavone i've opened the bug report https://github.com/VCVRack/Rack/issues/479

Does the problem may lay on the pcm sample rate? Can rtaudio handle it?

    $ arecord -f S24_3LE -c 2 -r 192000 -D hw:1,0 -d 5 test.wav
    Recording WAVE 'test.wav' : Signed 24 bit Little Endian in 3bytes, Rate 192000 Hz, Stereo

**Comment by radarsat1** [comment]
Don't know if it's a driver issue or not, but I'd definitely try to eliminate that possibility.  @spktkpkt I assume your last message has something to do with testing ALSA independently but I have no idea what you are trying to say.  Does arecord work or not?  If so it seems important to test a duplex situation.  Is there an ALSA-only duplex program like arecord that you could try?  Maybe this [test_latency.c](http://git.alsa-project.org/?p=alsa-lib.git;a=blob;f=test/latency.c;)?

Previous messages here mention 44100, have you tried 48000, 192000.. with rtaudio?  with arecord?

Another idea, have you tried using ALSA through Jack?

**Comment by spktkpkt** [comment]
@radarsat1 Yes, sorry, it means it works with 24bit. If I try with 16bit, arecord gives me a error. I got some other hints in the VCV bug report and will try them ASAP. I'll report back. Thanks so far.

**Comment by MarcBoule** [comment]
@radarsat1 Thanks very much for the followup, I'm quite eager to try `test_latency.c` as you suggested (which is a very good idea, to help see if it's an alsa issue or an RtAudio issue). I also had the idea for testing another application that uses alsa in duplex (tried Audigy but didn't manage to make it do duplex), so great suggestions all around. I'll report back after work! :+1:

**Comment by MarcBoule** [comment]
I got Jack working in a simple duplex test and everything works well:
![screenshot](https://user-images.githubusercontent.com/33604154/33585988-0d12aa3e-d935-11e7-9822-ccf3dbda2c39.png)

Regarding the other test I tried to run as suggested by @radarsat1, 
unfortunately I can't get `test_latency.c` from alsa-lib to run. Here's the output:

```
./latency -P hw:1,0 -C hw:1,0 -r 44100 -m 128 -M 128 -p -s 1
!!!Scheduler set to Round Robin with priority 99 FAILED!!!
Playback device is hw:1,0
Capture device is hw:1,0
Parameters are 44100Hz, S16_LE, 2 channels, non-blocking mode
Poll mode: yes
Loop limit is 44100 frames, minimum latency = 128, maximum latency = 128
ALSA lib conf.c:3544:(snd_config_hooks_call) Cannot open shared library libasound_module_conf_pulse.so
ALSA lib pcm.c:2501:(snd_pcm_open_noupdate) Unknown PCM hw:1,0
Playback open error: No such file or directory
```

The file `libasound_module_conf_pulse.so` is indeed locate here:
`ls /usr/lib/x86_64-linux-gnu/alsa-lib`, and I added this location to my $PATH variable. The alsa-lib package is not very user friendly to a non expert linux user like myself, so not sure what I'm doing wrong. I untared the package, ran `./configre` and [sudo] `make install` as they suggest. Then, in the test/ subfolder I ran `make latency`. Any suggestions would be greatly appreciated.

**Comment by radarsat1** [comment]
Ah, try adding it to LD_LIBRARY_PATH instead of PATH.  You got as far as compiling and running it, that's already pretty good for a "non expert linux user" ;)  Thanks for helping to debug!

I am worried that it is linking to an ALSA "PulseAudio" backend though that might not be representative.  You might want to pull test_latency.c out of the alsa repository and compile it manually against the version of ALSA actually installed on your system.

Anyway, for me, if Jack is working over ALSA with no problems, then that's an indication that there could be something wrong in the ALSA backend.  I can't tell you for sure.. can you list a few of the configurations that work/don't work?  It could be something as simple as an off-by-one error in a buffer calculation, but it's hard to tell without being able to test easily.

**Comment by spktkpkt** [comment]
@MarcBoule Looks like you own the UR22 MK2? I own the first generation. I don't know if they changed the chip(s) or just add the DC input and Class Compliance to it.

@radarsat1 Here is the output from the latency test.

    alsa-lib/test/latency -P hw:1,0 -C hw:1,0 -r 48000 -m 128 -M 128 -p -s 1 -f S24_3LE -r 192000
    Scheduler set to Round Robin with priority 99...
    Playback device is hw:1,0
    Capture device is hw:1,0
    Parameters are 192000Hz, S24_3LE, 2 channels, non-blocking mode
    Poll mode: yes
    Loop limit is 192000 frames, minimum latency = 128, maximum latency = 128
    Hardware PCM card 1 'Steinberg UR22' device 0 subdevice 0
    Its setup is:
      stream       : PLAYBACK
      access       : RW_INTERLEAVED
      format       : S24_3LE
      subformat    : STD
      channels     : 2
      rate         : 192000
      exact rate   : 192000 (192000/1)
      msbits       : 24
      buffer_size  : 128
      period_size  : 64
      period_time  : 333
      tstamp_mode  : NONE
      tstamp_type  : MONOTONIC
      period_step  : 1
      avail_min    : 64
      period_event : 0
      start_threshold  : 2147483647
      stop_threshold   : 128
      silence_threshold: 0
      silence_size : 0
      boundary     : 4611686018427387904
      appl_ptr     : 0
      hw_ptr       : 0
    Hardware PCM card 1 'Steinberg UR22' device 0 subdevice 0
    Its setup is:
      stream       : CAPTURE
      access       : RW_INTERLEAVED
      format       : S24_3LE
      subformat    : STD
      channels     : 2
      rate         : 192000
      exact rate   : 192000 (192000/1)
      msbits       : 24
      buffer_size  : 128
      period_size  : 64
      period_time  : 333
      tstamp_mode  : NONE
      tstamp_type  : MONOTONIC
      period_step  : 1
      avail_min    : 64
      period_event : 0
      start_threshold  : 2147483647
      stop_threshold   : 128
      silence_threshold: 0
      silence_size : 0
      boundary     : 4611686018427387904
      appl_ptr     : 0
      hw_ptr       : 0
    Trying latency 128 frames, 666.667us, 0.666667ms (1500.0000Hz)
    Success
    Playback:
    *** frames = 192128 ***
      state       : RUNNING
      trigger_time: 3059.790387726
      tstamp      : 0.000000
      delay       : 200
      avail       : 0
      avail_max   : 80
    Capture:
    *** frames = 192000 ***
      state       : RUNNING
      trigger_time: 3059.790117397
      tstamp      : 0.000000
      delay       : 47
      avail       : 47
      avail_max   : 112
    Maximum read: 64 frames
    Maximum read latency: 333.333us, 0.333333ms (3000.0000Hz)
    Hardware sync
    Playback time = 3059.790117, Record time = 3059.790117, diff = 0

Looks very decent as far as i can interpret the output. It faild with 64 frames (min/max latency), but 128 frames is good.

**Comment by spktkpkt** [comment]
@radarsat1 Now it works for me. I had to recompile rtaudio with alsa support. Thank you for your help and sorry for the inconvenience. I don't know if it also works for @MarcBoule, i guess he will also update the issue soon.

**Comment by MarcBoule** [comment]
Indeed, I will try the LD_LIBRARY_PATH suggestion for running the latency test when I get back home (you are correct that I have the UR22mk2, hope they didn't break anything in the newer version!). Not sure what you mean when you say that you had to compile rtaudio with alsa support, in VCV Rack this is done automatically... perhaps you mean in the alsa-lib package for testing, as referenced above? In any case, happy to see that all is good on your end! 

P.S. in your command for invoking the latency test, you had two switches that specify the frequency :-)

alsa-lib/test/latency -P hw:1,0 -C hw:1,0 **-r 48000** -m 128 -M 128 -p -s 1 -f S24_3LE **-r 192000**

**Comment by spktkpkt** [comment]
You're right, didn't noticed the doubles.

No, I really had to recompile the rtaudio libs for VCRack, see here
https://github.com/VCVRack/Rack/issues/479#issuecomment-349413411

**Comment by MarcBoule** [comment]
With the LD_LIBRARY_PATH correctly set, I was able to run the `latency.c` test. I had to use -f S32_LE. To my surprise, it says it doesn't support S16_LE, nor FLOAT_LE, which is quite standard, and leaves me wondering how a supposed USB audio interface is not able to handle those formats. Anyhow, here are my results (in the tests below, latency is from alsa-lib, and duplex is from rtaudio)

`./latency -P hw:1,0 -C hw:1,0 -r 44100 -m 128 -M 128 -p -s 10 -f S32_LE`
result: **static in the audio**

`./duplex 2 44100 4 4 0 0`      [bufferFrames = 128 in the code]
result: **static in the audio**

`./latency -P hw:1,0 -C hw:1,0 -r 44100 -m 16384 -M 16384 -p -s 10 -f S32_LE`
result: **poping/choppiness in the audio**  (2 to 3 pops per second approx)

`./duplex 2 44100 4 4 0 0`      [bufferFrames = 16384 in the code]
result: **poping/choppiness in the audio**  (2 to 3 pops per second approx)

Looks pretty bad I guess, since trouble is not in rtaudio. @radarsat1, your conjecture that the problem may be in the alsa backend would seem logical, and also that a buffer endpoint or startpoint might have a small offset. No idea how to proceed from here, and since the problem is not with rtaudio, I would understand if you don't spend further time on this. Conversely, if the problem interests you, I'm definitely in for more testing. I can send sound samples if ever you want to hear what it sounds like.

One more indication something is not going well in alsa with my UR22mkii: running `alsamixer -c1` leaves the screen with only one message (as opposed to the multiple mixer controls we normally see).
![image](https://user-images.githubusercontent.com/33604154/33639433-3c8ca458-d9f9-11e7-867c-fdb8b98dc3eb.png)

More outpus below for reference.

```
aplay -l**** List of PLAYBACK Hardware Devices ****
card 0: PCH [HDA Intel PCH], device 0: ALC3202 Analog [ALC3202 Analog]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: PCH [HDA Intel PCH], device 3: HDMI 0 [HDMI 0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: PCH [HDA Intel PCH], device 7: HDMI 1 [HDMI 1]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: PCH [HDA Intel PCH], device 8: HDMI 2 [HDMI 2]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 1: UR22mkII [Steinberg UR22mkII], device 0: USB Audio [USB Audio]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
```


```
./latency -P hw:1,0 -C hw:1,0 -r 44100 -m 128 -M 128 -p -s 10 -f S32_LE
!!!Scheduler set to Round Robin with priority 99 FAILED!!!
Playback device is hw:1,0
Capture device is hw:1,0
Parameters are 44100Hz, S32_LE, 2 channels, non-blocking mode
Poll mode: yes
Loop limit is 441000 frames, minimum latency = 128, maximum latency = 128
Hardware PCM card 1 'Steinberg UR22mkII' device 0 subdevice 0
Its setup is:
  stream       : PLAYBACK
  access       : RW_INTERLEAVED
  format       : S32_LE
  subformat    : STD
  channels     : 2
  rate         : 44100
  exact rate   : 44100 (44100/1)
  msbits       : 32
  buffer_size  : 128
  period_size  : 64
  period_time  : 1451
  tstamp_mode  : NONE
  tstamp_type  : MONOTONIC
  period_step  : 1
  avail_min    : 64
  period_event : 0
  start_threshold  : 2147483647
  stop_threshold   : 128
  silence_threshold: 0
  silence_size : 0
  boundary     : 4611686018427387904
  appl_ptr     : 0
  hw_ptr       : 0
Hardware PCM card 1 'Steinberg UR22mkII' device 0 subdevice 0
Its setup is:
  stream       : CAPTURE
  access       : RW_INTERLEAVED
  format       : S32_LE
  subformat    : STD
  channels     : 2
  rate         : 44100
  exact rate   : 44100 (44100/1)
  msbits       : 32
  buffer_size  : 128
  period_size  : 64
  period_time  : 1451
  tstamp_mode  : NONE
  tstamp_type  : MONOTONIC
  period_step  : 1
  avail_min    : 64
  period_event : 0
  start_threshold  : 2147483647
  stop_threshold   : 128
  silence_threshold: 0
  silence_size : 0
  boundary     : 4611686018427387904
  appl_ptr     : 0
  hw_ptr       : 0
Trying latency 128 frames, 2902.494us, 2.902494ms (344.5312Hz)
Success
Playback:
*** frames = 441152 ***
  state       : RUNNING
  trigger_time: 3092.067443
  tstamp      : 0.000000
  delay       : 152
  avail       : 4
  avail_max   : 69
Capture:
*** frames = 441024 ***
  state       : RUNNING
  trigger_time: 3092.066539
  tstamp      : 0.000000
  delay       : 5
  avail       : 5
  avail_max   : 86
Maximum read: 64 frames
Maximum read latency: 1451.247us, 1.451247ms (689.0625Hz)
Playback time = 3092.66540, Record time = 3092.66539, diff = 1
```


```
$ aplay -D hw:1 test.wav
Playing WAVE 'test.wav' : Signed 16 bit Little Endian, Rate 44100 Hz, Stereo
aplay: set_params:1233: Sample format non available
Available formats:
- S32_LE
```

**Comment by spktkpkt** [comment]
@MarcBoule the last output from latency test looks good to me. AFAIK it is normal that alsamixer do not show faders for volumes etc., because it is all managed by the hardware. Here are the commands which i have used to build VCVRack (this machine has 4 CPU cores, you might like to adjust the `-j` option on the `make` commands):

    sudo apt-get install git gcc make cmake tar unzip libglfw3-dev libglew-dev libzip-dev librtmidi-dev librtaudio-dev libgtk2.0-dev libjack-dev automake libtool
    
    git clone https://github.com/VCVRack/Rack.git
    
    cd Rack
    
    git submodule update --init --recursive
    
    make -j5 dep
    
    cd dep/rtaudio
    
    ./autogen.sh
    
    ./configure --with-jack --with-alsa
    
    make -j5
    
    cp -a .libs/librtaudio.so* ../lib/
    
    cd ../../
    
    make -j5
    
    make run

You can then copy the binary versions of the plugins into the *plugins* folder or build them too, if you like. I hope the `apt-get install ...` command includes all the dependencies. I have tested it 10 minutes ago on my other machine, which runs Ubuntu 16.04 LTS with a full GNOME3 desktop environment and also PulseAudio. No problems at all. Hope that helps.

**Comment by MarcBoule** [comment]
Other than the -j5 option, that is exactly the build sequence I am doing. Still no devices listed when I choose Jack in Rack. I'll keep poking around. If I can't get it working I'll post a separate issue. Thanks for all the discussion!

**Comment by spktkpkt** [comment]
Most welcome @MarcBoule! Just one last little hint. Make sure you have startet Jack (qjackctl) before starting VCVRack.

**Comment by radarsat1** [comment]
Hi, not sure what to propose for further tests, but I would say that if you can reproduce the problem with an ALSA-only example such as the latency test, which appears to be the case, then it's probably time to take the question to the ALSA users or developers list.  http://www.alsa-project.org/main/index.php/Mailing-lists

Probably someone there knows better how to deal with the specific hardware.  Who knows maybe it is even a known bug.

**Comment by MarcBoule** [comment]
Ok, will do. Thanks again for your time. Closing the issue.

**Comment by JonathanDotExe** [comment]
Sorry for commenting on this but I have exactly the same issue (and I also have the exact same interface) and I wanted to ask if you were able to solve your problem @MarcBoule .

Thanks in advance

**Comment by JonathanDotExe** [comment]
Ok. Thank you for your fast answer anyways :D

**Comment by wlaub** [comment]
I may be seeing a similar issue with my ES-8 - works fine in half duplex, but in full duplex shows regular errors in the output spaced roughly a buffer-width apart. The below example was taken in a VCV loopback test, but I was able to reproduce with the rtaudio playsaw test by modifying it to open the stream in full duplex.
![loopback10](https://user-images.githubusercontent.com/7267951/128072283-9646fd54-a9fa-4f1c-a5b2-d5a8c01b62f9.png)
I didn't have a whole lot of luck with the alsa-lib latency test, as it requires that the device have an equal number of input and output channels, which is not true of the ES-8.
While looking through the the rtaudio alsa code, I tried commenting out the [snd_pcm_link lines](https://github.com/thestk/rtaudio/blob/master/RtAudio.cpp#L7954), and that seems to have resolved the issue, though I don't know why.
